### ISSUE TITLE
Parallelization of Matrix Partitioning in Construction of Hierarchical Matrices using Task Parallel Languages

### ISSUE THREAD

**Issue by delta-leader** [rioyokotalab/fmm-summary#105]
## Single sentence description (like a tweet)
This paper tries to parallelize the cluster tree and block cluster tree creation during H-matrix construction.

### Link
https://www.jstage.jst.go.jp/article/ipsjjip/27/0/27_840/_article/-char/ja/

### Author and organization
Zhengyang Bai (Kyoto University), Tasuku Hiraishi (Tokyo University), Hiroshi Nakashima (Kyoto University), Akihiro Ida (University of Tokyo), Masahiro Yasugi (Kyushu Institute of Technology)

### Year/publication/conference/workshop
2019, Journal of Information Processing

## Abstract
A hierarchical matrix (H-matrix) is an approximated form that represents N × N correlations of N objects. H-matrix construction is achieved by partitioning a matrix into submatrices, followed by calculating the element values of these submatrices. This paper proposes implementations of matrix partitioning using task parallel languages, Cilk Plus and Tascell. Matrix partitioning is divided into two steps: cluster tree construction by dividing objects into clusters hierarchically, and block cluster tree construction by observing all cluster pairs at the same level of the cluster tree that satisfy an admissibility condition. As the two types of trees constructed and traversed in these steps are unpredictably unbalanced, it is expected that we can efficiently parallelize both these steps using task parallel languages. To obtain sufficient parallelism in the cluster tree construction, we not only execute recursive calls in parallel but also parallelize the inside of each recursive step. For the block cluster tree construction, we assigned each worker its own space so that the workers can store the cluster pairs without using locks. As a result, compared to a sequential implementation in C, we achieved up to an 11.5-fold speedup using Cilk Plus and a 12.6-fold speedup by Tascell for the cluster tree construction. For the block cluster tree construction, up to a 37.7-fold speedup by Cilk Plus and a 38.8-fold speedup using Tascell are achieved. In regard to the entire process of matrix partitioning, we achieved up to a 12.2-fold speedup by Cilk Plus and a 14.5-fold speedup using Tascell. All experiments were executed on two 18-core Xeon processors with real datasets to generate coefficient matrices used in the surface charge method.

## Innovation/contribution. What is new in this paper?
Parallelization of H-matrix creation usually focuses on the filling operation, an area where considerable speed-ups have been achieved in recent years. However, the actual tree creation remains serial in such codes and thus does not scale to the number of nodes. This might become a bottleneck in future implementations and hence this paper investigates ways to parallelize both the cluster tree and the block cluster tree creation during H-matrix setup.

## Short description of method used
For both trees, task parallelism is used applied to the recursive function calls used in the serial implementation. However, this is not able to reduce the critical path during the construction of the trees and thus further parallelization methods are applied during the cluster tree creation (i.e. parallel determination of min/max values, parallel sorting), which is the more complex task.
For the purpose of the implementation, the two task-parallel languages Cilk+ and Tascell are used, due to the authors claiming that lightweight task-parallelism is not available in OpenMP.

## Experiment and result
Speed-up for both languages shows a similar trend for all tested input datasets:
![image](https://user-images.githubusercontent.com/55976437/102975389-553b3100-4543-11eb-9199-db3623c5ecdc.png)
As can be seen in the figure above, good speed-up can be achieved for a low number of cores, with the performance gain steadily decreasing and ultimately reaching a plateau as the number of workers increases. Highest speed-ups observed are ~12 for the cluster tree construction and ~38 for the block cluster tree creation, while the overall trends are identical for both cases. It can be inferred that the cluster tree creation is the bottleneck out of those two tasks, even though more fain grained parallelism has already been applied to this task. 
A comparison to the OpenMP parallel for construct shows the advantages of those refined techniques used during the cluster tree creation:
![image](https://user-images.githubusercontent.com/55976437/102976035-4dc85780-4544-11eb-8a6c-f1871747f4a7.png)

## What is discussed in this paper (discussion section)? What is the future work?
Only the potential for further improving the cluster tree construction by optimizing the locality of memory access and load balancing are discussed. Future work includes the extension of the proposed method for distributed memory systems.

## Your comments
I don't have much to comment since I am unfamiliar with botch Cilk+ and Tascell, however, as far as I know task parallelism is available in OpenMP and t do not see any reason why performance should be worse using that language. A comparison of these implementations to an OpenMp task-based parallel code would be more interesting than parallel for construct used in this paper.
Furthermore the authors seem to be using their own variant of a parallel quick-sort, raising the question whether or not an well established parallel sorting algorithm might be beneficial to the implementation.

## Paper you should read next
As fas as I know this is the first attempt on parallelizing the cluster and block cluster tree creation during H-matrix construction, and as such, no further reading is available.
